qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/datamodel/JavaDataModelCache.java

```diff
@@ -61,8 +62,74 @@ public JavaDataModelCache(QuteProjectRegistry projectRegistry) {
 		return projectRegistry.getJavaDefinition(params);
 	}
 
-	public List<ValueResolver> getResolversFor(ResolvedJavaTypeInfo javaType) {
-		return valueResolversRegistry.getResolversFor(javaType);
+	public List<ValueResolver> getResolversFor(ResolvedJavaTypeInfo javaType, String projectUri) {
+		List<ValueResolver> matches = new ArrayList<>();
+		for (ValueResolver resolver : valueResolversRegistry.getResolvers()) {
+			if (matchResolver(javaType, resolver, projectUri)) {
+				matches.add(resolver);
+			}
+		}
+
+		List<ValueResolver> allResolvers = getValueResolvers(projectUri).getNow(null);
+		if (allResolvers != null) {
+			for (ValueResolver resolver : allResolvers) {
+				if (resolver.getNamespace() == null && matchResolver(javaType, resolver, projectUri)) {
+					matches.add(resolver);
+				}
+			}
+		}
+		return matches;
+	}
+
+	/**
+	 * Returns true if the given java type match the value resolver (if the type
+	 * matches the first parameter of the value resolver method) and false
+	 * otherwise.
+	 * 
+	 * @param javaType   the java type.
+	 * @param resolver   the value resolver.
+	 * @param projectUri the project Uri.
+	 * 
+	 * @return true if the given java type match the value resolver (if the type
+	 *         matches the first parameter of the value resolver method) and false
+	 *         otherwise.
+	 */
+	private boolean matchResolver(ResolvedJavaTypeInfo javaType, ValueResolver resolver, String projectUri) {
+		// Example with following signature:
+		// "orEmpty(arg : java.util.List<T>) : java.lang.Iterable<T>"
+		JavaParameterInfo parameter = resolver.getParameterAt(0); // arg : java.util.List<T>
+		if (parameter == null) {
+			return false;
+		}
+		if (parameter.getJavaType().isSingleGenericType()) {
+			// <T>
+			return true;
+		}
+		String parameterType = parameter.getJavaType().getName();
+		String resolvedTypeName = javaType.getName();
+		if (parameterType.equals(resolvedTypeName)) {
+			return true;
+		}
+		if (javaType.getExtendedTypes() != null) {
+			for (String extendedType : javaType.getExtendedTypes()) {
+				if (parameterType.equals(extendedType)) {
+					return true;
+				}
+			}
+		}
+		if (!javaType.getTypeParameters().isEmpty()) {
+			ResolvedJavaTypeInfo result = resolveJavaType(resolvedTypeName, projectUri).getNow(null);
+			if (result != null) {

```

Comment:
I believe Java supports short circuiting, so we can do `if (result != null && result.getExtendedTypes() != null) {`

Reply:
good catch, fixed